core/metrics-spi/src/main/java/software/amazon/awssdk/metrics/provider/MetricConfigurationProviderChain.java

```diff
@@ -0,0 +1,88 @@
+/*
+ * Copyright 2010-2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.awssdk.metrics.provider;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Set;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import software.amazon.awssdk.annotations.SdkProtectedApi;
+import software.amazon.awssdk.metrics.MetricCategory;
+
+/**
+ * Composite {@link MetricConfigurationProvider} that lets you specify a list of providers to get metric configuration.
+ *
+ * This chain decides the {@link MetricConfigurationProvider} to use at object construction time and cannot be changed
+ * later. The resolution policy is as follows:
+ *
+ * 1) Use the first provider in the chain that has returns true in {@link MetricConfigurationProvider#enabled()} method.
+ *
+ * 2) If none of the providers has metrics enabled as mentioned in above step, then use the first provider in the chain.
+ */
+@SdkProtectedApi
+public class MetricConfigurationProviderChain implements MetricConfigurationProvider {
+
+    private static final Logger log = LoggerFactory.getLogger(MetricConfigurationProviderChain.class);
+
+    private final List<MetricConfigurationProvider> providers;
```

Comment:
Why do we need this?

Reply:
If customer wants to control the feature via different providers like DynamoDB, system property, jmx etc.

Reply:
But we do we need to store the providers and not just the resolved provider(s)?